tests/structured_conf/test_structured_basic.py

```diff
@@ -153,6 +160,10 @@ def test_merge_structured_onto_dict_nested(self, class_type: str) -> None:
         assert c1 == {"user": {"name": 7}}
         # type of name becomes str
         assert c2 == {"user": {"name": "7", "age": "???"}}
+        assert isinstance(c2, DictConfig)
+        c2_user = c2._get_node("user")
+        assert c2_user is not None
+        assert c2_user._metadata.ref_type == module.User
```

Comment:
```
get_ref_type(c2, "user") == module.User
```

Reply:
Done in 9fabb0e

Although I'm a bit worried because I had to write it:
```python
assert get_ref_type(c2, "user") == Optional[module.User]
```
and this made me realize that if `dest` is optional, then it will remain optional even if `src` isn't. Although this could be changed, I'm not sure it would be a good idea (it's not like for types, where `Any` is a hint that the config was untyped: here we can't necessarilly assume that `optional == True` means that it was set by default to this value)

(I think I mentioned it before, but I'll say it again: given all the edge cases and weird things that might happen, I'm still not convinced that this whole thing is a good idea in the end, even if I kickstarted it)

Reply:
Yes.
I think if we can pull it off it's worth it.
The semantics we are going for is pretty clean:

During merge:
1. missing fields are not overriding the value of non-missing fields (regardless of Structured Configs or not).
2. typed missing fields are replacing the type of untyped fields and are validating it.

I think the simplifying effects from this are worth it even if it's complicated to implement in practice.

Reply:
> Done in [9fabb0e](https://github.com/omry/omegaconf/commit/9fabb0e3bcaef80d9fd4f859aae005b24ff1739f)
> 
> Although I'm a bit worried because I had to write it:
> 
> ```python
> assert get_ref_type(c2, "user") == Optional[module.User]
> ```
> 
> and this made me realize that if `dest` is optional, then it will remain optional even if `src` isn't. Although this could be changed, I'm not sure it would be a good idea (it's not like for types, where `Any` is a hint that the config was untyped: here we can't necessarilly assume that `optional == True` means that it was set by default to this value)

So see, I think this just found a bug.
By default, fields without a type are optional.
In this case we are replacing the type, so we should also transfer the optional bit to dest.

Reply:
> In this case we are replacing the type, so we should also transfer the optional bit to dest.

Fix in 786577a: I think it's necessary but not sure if it's sufficient, looking into potential other problems now

Reply:
Yeah, I got the same fix while playing with it here:
https://github.com/omry/omegaconf/pull/467/commits/c3edf17b159c147a880d5c71c3d5f6a688ed31a0

Reply:
> Yeah, I got the same fix while playing with it here:
> [c3edf17](https://github.com/omry/omegaconf/commit/c3edf17b159c147a880d5c71c3d5f6a688ed31a0)

Cool, resolving this conversation then